File: go/pkg/zerotier/address.go
/*
 * Copyright (c)2013-2020 ZeroTier, Inc.
 *
 * Use of this software is governed by the Business Source License included
 * in the LICENSE.TXT file in the project's root directory.
 *
 * Change Date: 2024-01-01
 *
 * On the date above, in accordance with the Business Source License, use
 * of this software will be governed by version 2.0 of the Apache License.
 */
/****/

package zerotier

import (
	"encoding/json"
	"fmt"
	"strconv"
)

// Address represents a 40-bit short ZeroTier address
type Address uint64

// NewAddressFromString parses a 10-digit ZeroTier address
func NewAddressFromString(s string) (Address, error) {
	if len(s) != 10 {
		return Address(0), ErrInvalidZeroTierAddress
	}
	a, err := strconv.ParseUint(s, 16, 64)
	return Address(a & 0xffffffffff), err
}

// NewAddressFromBytes reads a 5-byte 40-bit address.
func NewAddressFromBytes(b []byte) (Address, error) {
	if len(b) < 5 {
		return Address(0), ErrInvalidZeroTierAddress
	}
	return Address((uint64(b[0]) << 32) | (uint64(b[1]) << 24) | (uint64(b[2]) << 16) | (uint64(b[3]) << 8) | uint64(b[4])), nil
}

// Copy this address to a byte array, which must be 5 bytes in length or this will panic.
func (a Address) CopyTo(b []byte) {
	b[0] = byte(a >> 32)
	b[1] = byte(a >> 24)
	b[2] = byte(a >> 16)
	b[3] = byte(a >> 8)
	b[4] = byte(a)
}

// IsReserved returns true if this address is reserved and therefore is not valid for a real node.
func (a Address) IsReserved() bool { return a == 0 || (a>>32) == 0xff }

// String returns this address's 10-digit hex identifier
func (a Address) String() string {
	return fmt.Sprintf("%.10x", uint64(a))
}

// MarshalJSON marshals this Address as a string
func (a Address) MarshalJSON() ([]byte, error) {
	return []byte(fmt.Sprintf("\"%.10x\"", uint64(a))), nil
}

// UnmarshalJSON unmarshals this Address from a string
func (a *Address) UnmarshalJSON(j []byte) error {
	var s string
	err := json.Unmarshal(j, &s)
	if err != nil {
		return err
	}
	tmp, err := NewAddressFromString(s)
	*a = tmp
	return err
}

File: rust-zerotier-core/src/address.rs
/*
 * Copyright (c)2013-2021 ZeroTier, Inc.
 *
 * Use of this software is governed by the Business Source License included
 * in the LICENSE.TXT file in the project's root directory.
 *
 * Change Date: 2026-01-01
 *
 * On the date above, in accordance with the Business Source License, use
 * of this software will be governed by version 2.0 of the Apache License.
 */
/****/

#[derive(PartialEq, Eq, Clone, Copy, Ord, PartialOrd)]
pub struct Address(pub u64);

impl Default for Address {
    #[inline(always)]
    fn default() -> Address {
        Address(0)
    }
}

impl Address {
    #[inline(always)]
    fn to_bytes(&self) -> [u8; 5] {
        [(self.0 >> 32) as u8, (self.0 >> 24) as u8, (self.0 >> 16) as u8, (self.0 >> 8) as u8, self.0 as u8]
    }
}

impl From<&[u8]> for Address {
    #[inline(always)]
    fn from(bytes: &[u8]) -> Self {
        if bytes.len() >= 5 {
            Address(((bytes[0] as u64) << 32) | ((bytes[1] as u64) << 24) | ((bytes[2] as u64) << 16) | ((bytes[3] as u64) << 8) | (bytes[4] as u64))
        } else {
            Address(0)
        }
    }
}

impl ToString for Address {
    fn to_string(&self) -> String {
        format!("{:0>10x}", self.0)
    }
}

impl From<u64> for Address {
    #[inline(always)]
    fn from(i: u64) -> Self {
        Address(i)
    }
}

impl From<&str> for Address {
    #[inline(always)]
    fn from(s: &str) -> Self {
        Address(u64::from_str_radix(s, 16).unwrap_or(0))
    }
}

impl serde::Serialize for Address {
    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error> where S: serde::Serializer {
        if serializer.is_human_readable() {
            serializer.serialize_str(self.to_string().as_str())
        } else {
            let b = self.to_bytes();
            serializer.serialize_bytes(b.as_ref())
        }
    }
}

struct AddressVisitor;

impl<'de> serde::de::Visitor<'de> for AddressVisitor {
    type Value = Address;
    fn expecting(&self, formatter: &mut std::fmt::Formatter) -> std::fmt::Result { formatter.write_str("ZeroTier Address") }
    fn visit_str<E>(self, s: &str) -> Result<Self::Value, E> where E: serde::de::Error { Ok(Address::from(s)) }
    fn visit_bytes<E>(self, v: &[u8]) -> Result<Self::Value, E> where E: serde::de::Error { Ok(Address::from(v)) }
}

impl<'de> serde::Deserialize<'de> for Address {
    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error> where D: serde::Deserializer<'de> {
        if deserializer.is_human_readable() {
            deserializer.deserialize_str(AddressVisitor)
        } else {
            deserializer.deserialize_bytes(AddressVisitor)
        }
    }
}
